gym/envs/registration.py

```diff
@@ -9,17 +9,291 @@
 else:
     import importlib.metadata as metadata
 
+from collections import defaultdict
+from collections.abc import MutableMapping
+from operator import getitem
+
+from typing import Optional, Union, Dict, Set, Tuple, Generator
+
 from gym import error, logger
 
 # This format is true today, but it's *not* an official spec.
 # [username/](env-name)-v(version)    env-name is group 1, version is group 2
 #
 # 2016-10-31: We're experimentally expanding the environment ID format
 # to include an optional username.
-env_id_re = re.compile(r"^(?:[\w:-]+\/)?([\w:.-]+)-v(\d+)$")
+env_id_re: re.Pattern = re.compile(
+    r"^(?:(?P<namespace>[\w:-]+)\/)?(?P<name>[\w:.-]+)-v(?P<version>\d+)$"
+)
+
+
+def env_id_from_parts(namespace: str, name: str, version: Union[int, str]) -> str:
+    """
+    Construct the environment ID from the namespace, name, and version.
+    """
+    namespace = "" if namespace is None else f"{namespace}/"
+    return f"{namespace}{name}-v{version}"
+
 
 # Whitelist of plugins which can hook into the `gym.envs.internal` entry point.
-plugin_internal_whitelist = {"ale_py.gym"}
+plugin_internal_whitelist: Set[str] = {"ale_py.gym"}
+
+# The following is a map of environments which have been relocated
+# to a different namespace. This map is important when reporting
+# new versions of an environment outside of Gym.
+# This map should be removed eventually once users
+# are sufficiently aware of the environment relocations.
+# The value of the mapping is (namespace, package,).
+internal_env_namespace_relocation_map: Dict[str, Tuple[str, str]] = {
+    "Adventure": (
```

Comment:
This needs to be reworked to include the `-ram` environments. Perhaps create a new file with these relocations as this is getting quite cluttered.